File: anchor-reference-server/src/main/java/org/stellar/anchor/reference/service/FeeService.java
package org.stellar.anchor.reference.service;

import static org.stellar.anchor.util.MathHelper.decimal;

import java.math.BigDecimal;
import java.util.Optional;
import org.springframework.stereotype.Service;
import org.stellar.anchor.api.callback.GetFeeRequest;
import org.stellar.anchor.api.callback.GetFeeResponse;
import org.stellar.anchor.api.exception.BadRequestException;
import org.stellar.anchor.api.shared.Amount;
import org.stellar.anchor.reference.model.Customer;
import org.stellar.anchor.reference.repo.CustomerRepo;

@Service
public class FeeService {
  private final CustomerRepo customerRepo;
  BigDecimal feePercent = decimal("0.02"); // fixed 2% fee.
  BigDecimal feeFixed = decimal("0.1");

  FeeService(CustomerRepo customerRepo) {
    this.customerRepo = customerRepo;
  }

  public GetFeeResponse getFee(GetFeeRequest request) throws BadRequestException {
    if (request.getSendAsset() == null) {
      throw new BadRequestException("send_asset cannot be empty.");
    }

    if (request.getReceiveAsset() == null) {
      throw new BadRequestException("receive_asset cannot be empty.");
    }

    if (request.getClientId() == null) {
      throw new BadRequestException("client_id cannot be empty.");
    }

    if (request.getSendAmount() == null && request.getReceiveAmount() == null) {
      throw new BadRequestException("sender_amount or receiver_amount must be present.");
    }

    if (request.getSenderId() == null) {
      throw new BadRequestException("sender_id cannot be empty.");
    }
    Optional<Customer> maybeSender = customerRepo.findById(request.getSenderId());
    if (maybeSender.isEmpty()) {
      throw new BadRequestException("sender_id was not found.");
    }

    if (request.getReceiverId() == null) {
      throw new BadRequestException("receiver_id cannot be empty.");
    }
    Optional<Customer> maybeReceiver = customerRepo.findById(request.getReceiverId());
    if (maybeReceiver.isEmpty()) {
      throw new BadRequestException("receiver_id was not found.");
    }

    BigDecimal amount = decimal(request.getSendAmount());
    // fee = feeFixed + feePercent * sendAmount
    BigDecimal fee = amount.multiply(feePercent).add(feeFixed);
    return new GetFeeResponse(new Amount(String.valueOf(fee), request.getSendAsset()));
  }
}

File: kotlin-reference-server/src/main/kotlin/org/stellar/reference/callbacks/fee/FeeService.kt
package org.stellar.reference.callbacks.fee

import java.math.BigDecimal
import org.stellar.anchor.api.callback.GetFeeRequest
import org.stellar.anchor.api.callback.GetFeeResponse
import org.stellar.anchor.api.exception.UnprocessableEntityException
import org.stellar.anchor.api.shared.Amount
import org.stellar.reference.callbacks.BadRequestException
import org.stellar.reference.dao.CustomerRepository

class FeeService(private val customerRepository: CustomerRepository) {
  fun getFee(request: GetFeeRequest): GetFeeResponse {
    validateRequest(request)
    val amount = request.sendAmount?.let { BigDecimal(it) }
    val fee = amount?.multiply(BigDecimal("0.02"))?.add(BigDecimal("0.1"))

    return GetFeeResponse(Amount(fee!!.toString(), request.sendAsset))
  }

  private fun validateRequest(request: GetFeeRequest) {
    if (request.sendAsset == null) {
      throw BadRequestException("Send asset must be provided")
    }

    if (request.receiveAsset == null) {
      throw BadRequestException("Receive asset must be provided")
    }

    if (request.clientId == null) {
      throw BadRequestException("Client id must be provided")
    }

    if (request.sendAmount == null && request.receiveAmount == null) {
      throw BadRequestException("Either send or receive amount must be provided")
    }

    if (request.senderId == null) {
      throw BadRequestException("Sender id must be provided")
    } else {
      customerRepository.get(request.senderId)
        ?: throw UnprocessableEntityException("Sender ${request.senderId} not found")
    }

    if (request.receiverId == null) {
      throw BadRequestException("Receiver id must be provided")
    } else {
      customerRepository.get(request.receiverId)
        ?: throw UnprocessableEntityException("Receiver ${request.receiverId} not found")
    }
  }
}
